[SPARK-14163][CORE] SumEvaluator and countApprox cannot reliably handle RDDs of size 1

mtustin-handy opened the pull request:
## What changes were proposed in this pull request?

This special cases 0 and 1 counts to avoid passing 0 degrees of freedom. 
## How was this patch tested?

Tests run successfully. New test added. 
## Note:

This recreates #11982 which was closed to due to non-updated diff. @rxin @srowen Commented there. 
This also adds tests, reworks the code to perform the special casing (based on @srowen's comments), and adds equality machinery for BoundedDouble, as well as changing how it is transformed to string.

mtustin-handy committed:
Use confidence when count is 1.

Work with me in New York? https://www.handy.com/careers/73115?gh_jid=73115&gh_src=o5qcxn

mtustin-handy committed:
Set infinite confidence if count is 1 or otherwise confidence would be NaN.
Add test + equality methods for BoundedDouble

Work with me in NYC: https://www.handy.com/careers/73115?gh_jid=73115&gh_src=o5qcxn

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on core/src/main/scala/org/apache/spark/partial/BoundedDouble.scala:
```diff
@@ -21,5 +21,20 @@ package org.apache.spark.partial
  */
 class BoundedDouble(val mean: Double, val confidence: Double, val low: Double, val high: Double) {
-  override def toString(): String = "[%.3f, %.3f]".format(low, high)
+  override def toString(): String = "BoundedDouble(%.3f, %.3f, %.3f, %.3f)".format(mean, confidence, low, high)
```
I was going to say that this can become a case class to get this all for free, but it's part of the API. In theory that doesn't change anything, so you can try a case class and see if MiMa accepts it. Otherwise you can do this, but you'll need to fix up a few little style things. I'd not change the toString, hashCode might be too long a line, and indent on equals is too deep

rxin replied:
case classes bring extra concerns with binary compatibility (due to pattern matching). I'd minimize its use in public APIs.

srowen commented on core/src/main/scala/org/apache/spark/partial/SumEvaluator.scala:
```diff
@@ -40,30 +41,39 @@ private[spark] class SumEvaluator(totalOutputs: Int, confidence: Double)
+
+      // branch at this point because counter.count == 1 implies counter.sampleVariance == Nan
+      // and we don't want to ever return a bound of NaN
+      if (meanVar == Double.NaN || counter.count == 1) {
```
`NaN != NaN`; you'd have to use `Double.isNaN`

mtustin-handy replied:
Ah good catch

srowen commented on core/src/main/scala/org/apache/spark/partial/SumEvaluator.scala:
```diff
@@ -40,30 +41,39 @@ private[spark] class SumEvaluator(totalOutputs: Int, confidence: Double)
           val degreesOfFreedom = (counter.count - 1).toInt
           new TDistribution(degreesOfFreedom).inverseCumulativeProbability(1 - (1 - confidence) / 2)
+        } else {
+          throw new Exception("Counter.count <= 1; this should be impossible at this point")
```
This is unnecessary, as you say. Just remove this branch right?

mtustin-handy replied:
Unfortunately without the final branch confFactor resolves to Any type.
That last branch is necessary to keep it as double - either by returning a
dummy value or throwing an exception. I think an exception is safer because
it's more easily detected that it is no longer impossible.

I'll work on the tests tomorrow.

On Tuesday, March 29, 2016, Sean Owen <notifications@github.com> wrote:

> In core/src/main/scala/org/apache/spark/partial/SumEvaluator.scala
> <https://github.com/apache/spark/pull/12016#discussion_r57723805>:
>
> >            val degreesOfFreedom = (counter.count - 1).toInt
> >            new TDistribution(degreesOfFreedom).inverseCumulativeProbability(1 - (1 - confidence) / 2)
> > +        } else {
> > +          throw new Exception("Counter.count <= 1; this should be impossible at this point")
>
> This is unnecessary, as you say. Just remove this branch right?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/12016/files/3faecc4f18094686c843060a1e53b81b9e04e75d#r57723805>
>

-- 
Want to work at Handy? Check out our culture deck and open roles 
<http://www.handy.com/careers>
Latest news <http://www.handy.com/press> at Handy
Handy just raised $50m 
<http://venturebeat.com/2015/11/02/on-demand-home-service-handy-raises-50m-in-round-led-by-fidelity/> led 
by Fidelity

srowen replied:
... but I'm suggesting there be no third branch at all, like before. I understand the types.

mtustin-handy replied:
Right, but then you have to have an unconditional else as the final branch. This whole bug stems from that code being unsuitable for an unconditional else. With the code with three branches, and the final branch being an exception, you'll know if changes in the logic above have re-introduced the same error condition. With an unconditional else holding the call to TDistribution you won't.

srowen replied:
You've already handled the count=0 and count=1 cases earlier. Checking count > 1 doesn't do anything since it can't happen so having a branch for it is odd. Tests are how we catch regressions.

mtustin-handy replied:
I understand that the check does nothing for the computer but it makes it
easier to read. It's slightly better than a comment because it won't lie
around being incorrect and stale.

Nevertheless I can fix it up to your preference together the tests.

On Tuesday, March 29, 2016, Sean Owen <notifications@github.com> wrote:

> In core/src/main/scala/org/apache/spark/partial/SumEvaluator.scala
> <https://github.com/apache/spark/pull/12016#discussion_r57732511>:
>
> >            val degreesOfFreedom = (counter.count - 1).toInt
> >            new TDistribution(degreesOfFreedom).inverseCumulativeProbability(1 - (1 - confidence) / 2)
> > +        } else {
> > +          throw new Exception("Counter.count <= 1; this should be impossible at this point")
>
> You've already handled the count=0 and count=1 cases earlier. Checking
> count > 1 doesn't do anything since it can't happen so having a branch for
> it is odd. Tests are how we catch regressions.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/12016/files/3faecc4f18094686c843060a1e53b81b9e04e75d#r57732511>
>

-- 
Want to work at Handy? Check out our culture deck and open roles 
<http://www.handy.com/careers>
Latest news <http://www.handy.com/press> at Handy
Handy just raised $50m 
<http://venturebeat.com/2015/11/02/on-demand-home-service-handy-raises-50m-in-round-led-by-fidelity/> led 
by Fidelity

srowen replied:
I don't think it's easier to read since it's more code and suggests this branch can happen, when it should be impossible.

srowen commented on core/src/test/scala/org/apache/spark/partial/SumEvaluatorSuite.scala:
```diff
@@ -0,0 +1,49 @@
+
+class SumEvaluatorSuite extends SparkFunSuite with SharedSparkContext {
+
+  test("correct handling of count 1") {
```
While you're here throw in a basic test for count == 0 too, and ideally some normal-path case for completeness

mtustin-handy replied:
Done.

srowen commented on core/src/test/scala/org/apache/spark/partial/SumEvaluatorSuite.scala:
```diff
@@ -0,0 +1,49 @@
+    //execute
+    val res = evaluator.currentResult()
+    // Build version with known precisions for equality check
+    val round_res = new BoundedDouble(res.mean.round.toDouble, res.confidence, res.low, res.high)
```
I don't see a reason to round to the nearest integer. You can assert approximate equality in a test. But I think you can assert exact equality in this case since it's deterministic

mtustin-handy replied:
Unfortunately, the exact value being returned doesn't have an exact decimal representation. I'll think about the best course here.

srowen replied:
By that logic lots of things returning floating point values would be untestable, but they are. You are asserting that the value returned is the expected closest-possible double. This is the best course. This test asserts anything that rounds to the same int is fine, but that's unnecessarily coarse, and we have clearer syntax for it. If it's really called for, asserting approximate equality is better.

This should be a conceptually simple fix plus tests, so I'd prefer to head straight at the simple solution and wrap this up, or else it may be simpler to merge https://github.com/apache/spark/pull/11981 with a test instead.

mtustin-handy replied:
@srowen Update me then if you change your mind. In the mean time I'll make changes here.

mtustin-handy committed:
Update tests, reformat additions.

Also, work with me in NYC: https://www.handy.com/careers/73115?gh_jid=73115&gh_src=o5qcxn

srowen commented:
Jenkins retest this please

srowen commented on core/src/main/scala/org/apache/spark/partial/BoundedDouble.scala:
```diff
@@ -21,5 +21,23 @@ package org.apache.spark.partial
+  override def toString(): String =
+    "BoundedDouble(%.3f, %.3f, %.3f, %.3f)".format(mean, confidence, low, high)
+
+  override def hashCode:Int =
```
(This might fail style checks without a space after : )

srowen commented on core/src/test/scala/org/apache/spark/partial/SumEvaluatorSuite.scala:
```diff
@@ -0,0 +1,117 @@
+    //execute
+    val res = evaluator.currentResult()
+    // 38.0 - 7.1E-15 because that's how the maths shakes out
+    val target_mean =  38.0 - 7.1E-15
```
Looking good, now into nits. var names should be camelCase instead of with_underscores.

mtustin-handy replied:
My python is showing 😬

On Thursday, March 31, 2016, Sean Owen <notifications@github.com> wrote:

> In core/src/test/scala/org/apache/spark/partial/SumEvaluatorSuite.scala
> <https://github.com/apache/spark/pull/12016#discussion_r58143122>:
>
> > +class SumEvaluatorSuite extends SparkFunSuite with SharedSparkContext {
> > +
> > +  test("correct handling of count 1") {
> > +
> > +    //setup
> > +    val counter = new StatCounter(List(2.0))
> > +    // count of 10 because it's larger than 1,
> > +    // and 0.95 because that's the default
> > +    val evaluator = new SumEvaluator(10, 0.95)
> > +    // arbitrarily assign id 1
> > +    evaluator.merge(1, counter)
> > +
> > +    //execute
> > +    val res = evaluator.currentResult()
> > +    // 38.0 - 7.1E-15 because that's how the maths shakes out
> > +    val target_mean =  38.0 - 7.1E-15
>
> Looking good, now into nits. var names should be camelCase instead of
> with_underscores.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/12016/files/93bcd99b11e0c914053cb1cfbe37fe0df4ba1a38#r58143122>
>

-- 
Want to work at Handy? Check out our culture deck and open roles 
<http://www.handy.com/careers>
Latest news <http://www.handy.com/press> at Handy
Handy just raised $50m 
<http://venturebeat.com/2015/11/02/on-demand-home-service-handy-raises-50m-in-round-led-by-fidelity/> led 
by Fidelity

SparkQA commented:
**[Test build #54677 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54677/consoleFull)** for PR 12016 at commit [`93bcd99`](https://github.com/apache/spark/commit/93bcd99b11e0c914053cb1cfbe37fe0df4ba1a38).

srowen commented on core/src/test/scala/org/apache/spark/partial/SumEvaluatorSuite.scala:
```diff
@@ -0,0 +1,117 @@
+    assert(ancillary_res == new BoundedDouble(target_mean, 0.950, Double.NegativeInfinity, Double.PositiveInfinity))
+
+    // check that mean is within expected tolerance of expectation
+    assert(mean_diff < Double.MinPositiveValue)
```
This is almost like just asserting equality. If we're this sure about the result then I assume we're actually sure of exact equality, and so would be fine to assert that directly for clarity.

To assert approximate equality you can see the other tests that use syntax like `assert(a ~== b absTol 1e-7)`. Approximate equality assertions are likely to involve asserting some more possible variance than the smallest possible double. Where it's necessary I'd expect you need wider tolerance; where it's not necessary, it should just be an exact equality assertion.

mtustin-handy replied:
TIL.

Style errors are all fixed according to local scalastyle.

mtustin-handy replied:
And, these tests are much nicer now they're equality checks, if I do say so myself.

SparkQA commented:
**[Test build #54677 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54677/consoleFull)** for PR 12016 at commit [`93bcd99`](https://github.com/apache/spark/commit/93bcd99b11e0c914053cb1cfbe37fe0df4ba1a38).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54677/
Test FAILed.

mtustin-handy committed:
Fix style errors, use equality now exact values are known.

Work with me in NYC: https://www.handy.com/careers/73115?gh_jid=73115&gh_src=o5qcxn

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #54729 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54729/consoleFull)** for PR 12016 at commit [`5db9678`](https://github.com/apache/spark/commit/5db9678631d5f747ac383032841cd27cbe6831b5).

SparkQA commented:
**[Test build #54729 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54729/consoleFull)** for PR 12016 at commit [`5db9678`](https://github.com/apache/spark/commit/5db9678631d5f747ac383032841cd27cbe6831b5).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54729/
Test FAILed.

srowen commented on core/src/test/scala/org/apache/spark/partial/SumEvaluatorSuite.scala:
```diff
@@ -0,0 +1,107 @@
+  test("correct handling of NaN") {
+
+    //setup
+    val counter = new StatCounter(List(1,Double.NaN,2))
```
Close now. It doesn't like the whitespace at ends of lines, and commas everywhere need a spaceafter

mtustin-handy replied:
Yeah, for some reason `sbt scalastyle` isn't checking my test code. I'll try to retrigger jenkins in a second.

srowen replied:
Try `./dev/lint-scala`

mtustin-handy committed:
Fix scalastyle errors.

Work with me in NYC: https://www.handy.com/careers/73115?gh_jid=73115&gh_src=o5qcxn

mtustin-handy commented:
Jenkins retest this please

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #54777 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54777/consoleFull)** for PR 12016 at commit [`5e3c477`](https://github.com/apache/spark/commit/5e3c47762f79b89544360c383db10b3d77411109).

SparkQA commented:
**[Test build #54777 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54777/consoleFull)** for PR 12016 at commit [`5e3c477`](https://github.com/apache/spark/commit/5e3c47762f79b89544360c383db10b3d77411109).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54777/
Test FAILed.

mtustin-handy commented:
The only failure has been failing since this: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54771/.

I can pull from master and try again.

srowen commented:
It's unrelated, I've seen it in other tests.

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #54782 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54782/consoleFull)** for PR 12016 at commit [`5e3c477`](https://github.com/apache/spark/commit/5e3c47762f79b89544360c383db10b3d77411109).

SparkQA commented:
**[Test build #54782 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54782/consoleFull)** for PR 12016 at commit [`5e3c477`](https://github.com/apache/spark/commit/5e3c47762f79b89544360c383db10b3d77411109).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54782/
Test PASSed.

srowen commented on core/src/main/scala/org/apache/spark/partial/BoundedDouble.scala:
```diff
@@ -21,5 +21,23 @@ package org.apache.spark.partial
  */
 class BoundedDouble(val mean: Double, val confidence: Double, val low: Double, val high: Double) {
-  override def toString(): String = "[%.3f, %.3f]".format(low, high)
+  override def toString(): String =
```
OK, I think this is all good, except I think the `toString` should be left alone. I forgot to mention this. Not that I really expect anyone to depend on the format, but let's leave it since it's a public class.

mtustin-handy replied:
I definitely can put it back, but the previous toString was just weird - it
only printed the bounds. Anyway, I'll update this in a sec (to go back).
Let me know if you change your mind.

On Saturday, April 2, 2016, Sean Owen <notifications@github.com> wrote:

> In core/src/main/scala/org/apache/spark/partial/BoundedDouble.scala
> <https://github.com/apache/spark/pull/12016#discussion_r58301669>:
>
> > @@ -21,5 +21,23 @@ package org.apache.spark.partial
> >   * A Double value with error bars and associated confidence.
> >   */
> >  class BoundedDouble(val mean: Double, val confidence: Double, val low: Double, val high: Double) {
> > -  override def toString(): String = "[%.3f, %.3f]".format(low, high)
> > +  override def toString(): String =
>
> OK, I think this is all good, except I think the toString should be left
> alone. I forgot to mention this. Not that I really expect anyone to depend
> on the format, but let's leave it since it's a public class.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/12016/files/5e3c47762f79b89544360c383db10b3d77411109#r58301669>
>

-- 
Want to work at Handy? Check out our culture deck and open roles 
<http://www.handy.com/careers>
Latest news <http://www.handy.com/press> at Handy
Handy just raised $50m 
<http://venturebeat.com/2015/11/02/on-demand-home-service-handy-raises-50m-in-round-led-by-fidelity/> led 
by Fidelity

mtustin-handy replied:
Done.

mtustin-handy committed:
Put toString back to how it was (only print bounds).

Work with me at Handy in NYC: https://www.handy.com/careers/73115?gh_jid=73115&gh_src=o5qcxn

srowen commented:
Looks good, thank you. Let's give it one more spin

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #54788 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54788/consoleFull)** for PR 12016 at commit [`4040e0e`](https://github.com/apache/spark/commit/4040e0ec2421d5abe9b89785955e1e3d2215676e).

SparkQA commented:
**[Test build #54788 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54788/consoleFull)** for PR 12016 at commit [`4040e0e`](https://github.com/apache/spark/commit/4040e0ec2421d5abe9b89785955e1e3d2215676e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54788/
Test PASSed.

srowen commented:
Merged to master, thanks
